File: service/src/main/java/hu/bsstudio/robonaut/team/DefaultTeamService.java
package hu.bsstudio.robonaut.team;

import hu.bsstudio.robonaut.entity.TeamEntity;
import hu.bsstudio.robonaut.repository.TeamRepository;
import hu.bsstudio.robonaut.team.mapper.TeamModelEntityMapper;
import hu.bsstudio.robonaut.team.model.DetailedTeam;
import hu.bsstudio.robonaut.team.model.Team;
import java.util.Collections;
import lombok.AccessLevel;
import lombok.NonNull;
import lombok.RequiredArgsConstructor;
import lombok.Setter;
import reactor.core.publisher.Flux;
import reactor.core.publisher.Mono;

@RequiredArgsConstructor
public class DefaultTeamService implements TeamService {

    @NonNull
    private final TeamRepository teamRepository;

    @Setter(AccessLevel.PACKAGE)
    private TeamModelEntityMapper teamMapper = new TeamModelEntityMapper();

    @Override
    public Mono<DetailedTeam> addTeam(final Team team) {
        return Mono.just(team)
            .map(this::toEntity)
            .flatMap(teamRepository::insert)
            .map(teamMapper::toModel);
    }

    @Override
    public Mono<DetailedTeam> updateTeam(final Team team) {
        return Mono.just(team)
            .map(Team::getTeamId)
            .flatMap(teamRepository::findById)
            .map(entity -> updateBasicTeamInfo(entity, team))
            .flatMap(teamRepository::save)
            .map(teamMapper::toModel);
    }

    @Override
    public Flux<DetailedTeam> findAllTeam() {
        return teamRepository.findAll()
            .map(teamMapper::toModel);
    }

    private TeamEntity toEntity(final Team team) {
        return updateBasicTeamInfo(new TeamEntity(), team);
    }

    private TeamEntity updateBasicTeamInfo(final TeamEntity entity, final Team team) {
        entity.setTeamId(team.getTeamId());
        entity.setYear(team.getYear());
        entity.setTeamName(team.getTeamName());
        entity.setTeamMembers(team.getTeamMembers());
        entity.setTeamType(team.getTeamType());
        entity.setSpeedTimes(Collections.emptyList());
        return entity;
    }
}

File: service/src/main/kotlin/hu/bsstudio/robonaut/team/DefaultTeamService.kt
package hu.bsstudio.robonaut.team

import hu.bsstudio.robonaut.entity.ScoreEntity
import hu.bsstudio.robonaut.entity.TeamEntity
import hu.bsstudio.robonaut.repository.TeamRepository
import hu.bsstudio.robonaut.team.mapper.TeamModelEntityMapper
import hu.bsstudio.robonaut.team.model.DetailedTeam
import hu.bsstudio.robonaut.team.model.Team
import reactor.core.publisher.Flux
import reactor.core.publisher.Mono

class DefaultTeamService(
    private val teamRepository: TeamRepository,
    private val teamMapper: TeamModelEntityMapper = TeamModelEntityMapper()
) : TeamService {

    override fun addTeam(team: Team): Mono<DetailedTeam> {
        return Mono.just(team)
            .map(::toEntity)
            .flatMap(teamRepository::insert)
            .map(teamMapper::toModel)
    }

    override fun updateTeam(team: Team): Mono<DetailedTeam> {
        return Mono.just(team)
            .map(Team::teamId)
            .flatMap(teamRepository::findById)
            .map { updateBasicTeamInfo(it, team) }
            .flatMap(teamRepository::save)
            .map(teamMapper::toModel)
    }

    override fun updateTeam(detailedTeam: DetailedTeam): Mono<DetailedTeam> {
        return Mono.just(detailedTeam)
            .map(teamMapper::toEntity)
            .flatMap(teamRepository::save)
            .map(teamMapper::toModel)
    }

    override fun findAllTeam(): Flux<DetailedTeam> {
        return teamRepository.findAll()
            .map(teamMapper::toModel)
    }

    private fun toEntity(team: Team) = TeamEntity(
        teamId = team.teamId,
        year = team.year,
        teamName = team.teamName,
        teamMembers = team.teamMembers,
        teamType = team.teamType,
        speedTimes = emptyList(),
        score = ScoreEntity(),
        juniorScore = ScoreEntity(),
    )

    private fun updateBasicTeamInfo(entity: TeamEntity, team: Team): TeamEntity {
        entity.teamId = team.teamId
        entity.year = team.year
        entity.teamName = team.teamName
        entity.teamMembers = team.teamMembers
        entity.teamType = team.teamType
        entity.speedTimes = emptyList()
        return entity
    }
}
